presto-main/src/test/java/com/facebook/presto/operator/TestExchangeClient.java

```diff
@@ -329,6 +335,69 @@ public void testClose()
         assertStatus(clientStatus, location, "closed", "not scheduled");
     }
 
+    @Test
+    public void testInitialRequestLimit()
+    {
+        DataSize maxResponseSize = new DataSize(1, BYTE);
+        MockExchangeRequestProcessor processor = new MockExchangeRequestProcessor(maxResponseSize);
+        List<URI> locations = new ArrayList<>();
+        int initialRequestCount = 8;
+        int numLocations = initialRequestCount * 2;
+
+        // add pages
+        for (int ii = 0; ii < numLocations; ii++)
+        {
+            URI location = URI.create("http://localhost:" + (8080 + ii));
+            locations.add(location);
+
+            processor.addPage(location, createPage(1));
+            processor.addPage(location, createPage(2));
+            processor.addPage(location, createPage(3));
+
+            processor.setComplete(location);
+        }
+
+        try (ExchangeClient exchangeClient = new ExchangeClient(
+                new DataSize(8, BYTE),
+                maxResponseSize,
+                initialRequestCount,
+                1,
+                new Duration(1, MINUTES),
+                true,
+                new TestingHttpClient(processor, testingHttpClientExecutor),
+                scheduler,
+                new SimpleLocalMemoryContext(newSimpleAggregatedMemoryContext(), "test"),
+                pageBufferClientCallbackExecutor)) {
+
+            for (int ii = 0; ii < numLocations; ii++) {
+                exchangeClient.addLocation(locations.get(ii), TaskId.valueOf("taskid" + ii));
+            }
+            exchangeClient.noMoreLocations();
+            assertFalse(exchangeClient.isClosed());
+
+            long start = System.nanoTime();
+
+            // start fetching pages
+            exchangeClient.scheduleRequestIfNecessary();
+            // wait for a page to be fetched
+            do {
+                // there is no thread coordination here, so sleep is the best we can do
+                assertLessThan(Duration.nanosSince(start), new Duration(5, TimeUnit.SECONDS));
+                sleepUninterruptibly(100, MILLISECONDS);
+            }
+            while (exchangeClient.getStatus().getBufferedPages() < 8);
+
+            // client should have sent 8 requests for a single page
+            assertEquals(exchangeClient.getStatus().getBufferedPages(), 8);
```

Comment:
Since `addLocation` calls to the `scheduleRequestIfNecessary` the requests are scheduled not all at once. Because of that i have this gut feeling that there could be potential executions that will result in either more than 8 bytes buffered, or in less then 8 buffered and a kind of a terminal state. 

Could you please do something like `@Test(invocationCount = 10000)` just to make sure?

Reply:
Should the test just use a direct executor? Not seeing why it needs to be multi-threaded.

Reply:
FYI it does pass in a loop if you use threads. I changed it to direct executor just to make it predictable.